storage/store/results.go

```diff
@@ -54,107 +71,46 @@ func (r *ExecutionResults) byID(resultID flow.Identifier) (*flow.ExecutionResult
-					blockID, resultID, storedResultID, storage.ErrDataMismatch)
-			}
-
-			// if the result is the same, we don't need to index it again
-			return nil
-		}
-
-		// if the result is not indexed, we can index it
-	}
-
-	err := operation.IndexExecutionResult(w, blockID, resultID)
-	if err == nil {
-		return nil
-	}
-
-	return nil
-}
-
-func (r *ExecutionResults) Store(result *flow.ExecutionResult) error {
-	return r.db.WithReaderBatchWriter(func(rw storage.ReaderBatchWriter) error {
-		return r.store(rw, result)
-	})
+// BatchIndex indexes an execution result by block ID in a given batch
+// The caller must acquire [storage.LockIndexExecutionResult]
+// It returns [storage.ErrDataMismatch] if there is already an indexed result for the given blockID,
+// but it is different from the given resultID.
+func (r *ExecutionResults) BatchIndex(lctx lockctx.Proof, rw storage.ReaderBatchWriter, blockID flow.Identifier, resultID flow.Identifier) error {
+	return r.indexCache.PutWithLockTx(lctx, rw, blockID, resultID)
 }
 
+// BatchStore stores an execution result in a given batch
+// No error is expected during normal operation.
 func (r *ExecutionResults) BatchStore(result *flow.ExecutionResult, batch storage.ReaderBatchWriter) error {
 	return r.store(batch, result)
 }
 
-func (r *ExecutionResults) BatchIndex(blockID flow.Identifier, resultID flow.Identifier, batch storage.ReaderBatchWriter) error {
-	return operation.IndexExecutionResult(batch.Writer(), blockID, resultID)
-}
-
+// ByID retrieves an execution result by its ID. Returns `ErrNotFound` if `resultID` is unknown.
 func (r *ExecutionResults) ByID(resultID flow.Identifier) (*flow.ExecutionResult, error) {
 	return r.byID(resultID)
 }
 
-// Index indexes an execution result by block ID.
-// Note: this method call is not concurrent safe, because it checks if the different result is already indexed
-// by the same blockID, and if it is, it returns an error.
-// The caller needs to ensure that there is no concurrent call to this method with the same blockID.
-func (r *ExecutionResults) Index(blockID flow.Identifier, resultID flow.Identifier) error {
-	err := r.db.WithReaderBatchWriter(func(rw storage.ReaderBatchWriter) error {
-		return r.index(rw.Writer(), blockID, resultID, false)
-	})
-
-	if err != nil {
-		return fmt.Errorf("could not index execution result: %w", err)
-	}
-	return nil
-}
-
-func (r *ExecutionResults) ForceIndex(blockID flow.Identifier, resultID flow.Identifier) error {
-	err := r.db.WithReaderBatchWriter(func(rw storage.ReaderBatchWriter) error {
-		return r.index(rw.Writer(), blockID, resultID, true)
-	})
-
-	if err != nil {
-		return fmt.Errorf("could not index execution result: %w", err)
-	}
-	return nil
-}
-
+// ByBlockID retrieves an execution result by block ID.
+// It returns [storage.ErrNotFound] if `blockID` does not refer to a block executed by this node
```

Comment:
```suggestion
// It returns [storage.ErrNotFound] if `blockID` refers to a block which is unknown, or for which a trusted (sealed or executed by this node) execution result does not exist.
```
Nitpick: Since ANs use this API, "executed by this node" is not quite accurate.